Add plugin doc check in CI

JarvisG495 opened the pull request:
- Add plugin doc check in CI
- Fix a dead link
- Replace `tools/doc/plugin_doc_gen.py` with `make doc-gen` in contribution docs
<!-- Uncomment the following checklist WHEN AND ONLY WHEN you're adding a new plugin -->
<!--
- [ ] Add a test case for the new plugin
- [ ] Add a CHANGELOG entry for the new plugin
- [ ] Add a component id in [the main repo](https://github.com/apache/skywalking/blob/master/oap-server/server-bootstrap/src/main/resources/component-libraries.yml#L415)
- [ ] Add a logo in [the UI repo](https://github.com/apache/skywalking-rocketbot-ui/tree/master/src/views/components/topology/assets)
- [ ] Rebuild the `requirements.txt` by running `tools/env/build_requirements_(linux|windows).sh`
- [ ] Rebuild the `Plugins.md` documentation by running `tools/doc/plugin_doc_gen.py`
-->

kezhenxu94 commented on Makefile:
```diff
@@ -55,7 +55,16 @@ dev-fix: dev-setup
 
 doc-gen:
-	$(VENV)/python tools/doc/plugin_doc_gen.py
+	python3 -m tools.doc.plugin_doc_gen
```
can we keep using `$(VENV)/python`?

JarvisG495 replied:
Yes. I made this change only because CI environment does not set up venv and check-doc-gen needs doc-gen.
Maybe we can keep using `$(VENV)/python` in doc-gen and use `python3 -m tools.doc.plugin_doc_gen` instead of `make doc-gen` in check-doc-gen.
What do you think?

kezhenxu94 replied:
@JarvisG495 let's add a goal in Makefile

```makefile
$(VENV):
	python3 -m venv $(VENV_DIR)
```

And add dependent `$(VENV)` to all other goals that use `$(VENV)`, so that it `venv` can  be initialized automatically

kezhenxu94 commented on Makefile:
```diff
@@ -43,6 +46,7 @@ lint: clean
 
 # used in development
 dev-setup:
+	$(VENV)
```
This 👇 is the correct form of "dependent"

```suggestion
dev-setup: $(VENV)
```

JarvisG495 replied:
Oh, I see. Corrected.

kezhenxu94 reviewed (approved):
Thank you @JarvisG495

Superskyyy commented:
It seems like CI wouldn't pass, let's take a look

Superskyyy commented:
````
ERROR: for 2488337382_provider_1  Cannot start service provider: failed to create shim: OCI runtime create failed: container_linux.go:380: starting container process caused: exec: "sw-python": executable file not found in $PATH: unknown
```

The testing env is using the wrong Python environment, introduced by venv, let's fix it.

Superskyyy commented:
Very nice, LGTM, thank you.

Superskyyy commented:
Let's try move these lines to below license-and-lint, also add this job name as a prerequisite of further ones like what we did in license and lint, so we can fail fast if these basic checks fail.
https://github.com/apache/skywalking-python/blob/f430e46c7e8f1edba281f46d174d4dfd8ae35f0f/.github/workflows/CI.yaml#L101-L115

JarvisG495 commented:
> Let's try move these lines to below license-and-lint, also add this job name as a prerequisite of further ones like what we did in license and lint, so we can fail fast if these basic checks fail.
> 
> https://github.com/apache/skywalking-python/blob/f430e46c7e8f1edba281f46d174d4dfd8ae35f0f/.github/workflows/CI.yaml#L101-L115

np, and I think these should be moved to the part right below `changes`, since it can be skipped if no changes are found, right?

Superskyyy commented:
> > Let's try move these lines to below license-and-lint, also add this job name as a prerequisite of further ones like what we did in license and lint, so we can fail fast if these basic checks fail.
> 
> > 
> 
> > https://github.com/apache/skywalking-python/blob/f430e46c7e8f1edba281f46d174d4dfd8ae35f0f/.github/workflows/CI.yaml#L101-L115
> 
> 
> 
> np, and I think these should be moved to the part right below `changes`, since it can be skipped if no changes are found, right?

Imagine when you accidentally/ unknowingly change the plugin documentation, it will not trigger the changes job because it's in the docs folder, thus skipping the check, but this shouldn't happen right?

JarvisG495 commented:
> > > Let's try move these lines to below license-and-lint, also add this job name as a prerequisite of further ones like what we did in license and lint, so we can fail fast if these basic checks fail.
> > 
> > 
> > > 
> > 
> > 
> > > https://github.com/apache/skywalking-python/blob/f430e46c7e8f1edba281f46d174d4dfd8ae35f0f/.github/workflows/CI.yaml#L101-L115
> > 
> > 
> > np, and I think these should be moved to the part right below `changes`, since it can be skipped if no changes are found, right?
> 
> Imagine when you accidentally/ unknowingly change the plugin documentation, it will not trigger the changes job because it's in the docs folder, thus skipping the check, but this shouldn't happen right?

Make sense. I missed that case. I have updated `plugin-doc-check` location and also modified the exit codes in `CheckStatus` accordingly to follow the execution order.

Superskyyy commented:
Looks great.
